### PR TITLE
WIP: Support for new `graphql-js` subscriptions support

### PR DESCRIPTION
**Pull request by Urigo**
## Background
Based on `graphql-js` with support for `subscriptions` (https://github.com/graphql/graphql-js/pull/846) with `AsyncIterator`.

This PR is based on @mistic PR: https://github.com/apollographql/subscriptions-transport-ws/pull/108 and it will be merged along with his changes and the new package's API.

## Changes

The main changes are support for `subscribe` and `execute` directly from `graphql-js` package, and manage the subscription/query/mutation (s).

Also, transformation between the `AsyncIterator` returned from `graphql-js` to local implementation of `Observable` and publish the result over the socket.

## Usage Example

With the new `graphql-js` package, the usage with this transport is as follow:

```js
import { subscribe, execute } from 'graphql';
import { SubscriptionServer } from 'subscriptions-transport-ws';
import schema from './schema';

  new SubscriptionServer(
    {
      schema,
      subscribe,
      execute,
    });
```

### CODE REVIEW

File: src/server.ts [hunk 1]
```diff
@@ -30,40 +30,32 @@ export interface RequestMessage {
...
+                                       contextValue?: any,
+                                       variableValues?: { [key: string]: any },
+                                       operationName?: string,) => IObservable<ExecutionResult>;
+
+export type ExecuteFunction = (schema: GraphQLSchema,
+                               document: DocumentNode,
+                               rootValue?: any,
+                               contextValue?: any,
+                               variableValues?: { [key: string]: any },
+                               operationName?: string,) => Promise<ExecutionResult>;
```
Comment (DxCx) [inline src/server.ts]: Missing definition for SubscribeFunction

File: src/server.ts [hunk 2]
```diff
@@ -30,40 +30,32 @@ export interface RequestMessage {
...
-export type ExecuteFunction = (
-  schema: GraphQLSchema,
-  document: DocumentNode,
-  rootValue?: any,
-  contextValue?: any,
-  variableValues?: {[key: string]: any},
-  operationName?: string,
-) => Promise<ExecutionResult>;
-
-export interface Executor {
```
Comment (DxCx) [inline src/server.ts]: Executor shouldn't be remove, but subscribe should be added to it.

File: src/server.ts [hunk 3]
```diff
@@ -89,9 +81,8 @@ class ExecuteAdapters {
...
             rootValue?: any,
             contextValue?: any,
-            variableValues?: {[key: string]: any},
-            operationName?: string,
-    ) => ({
+            variableValues?: { [key: string]: any },
+            operationName?: string,) => ({
       subscribe: (observer) => {
         if (ExecuteAdapters.isASubscriptionRequest(document, operationName)) {
           observer.error(new Error('Subscriptions are not supported'));
```
Comment (DxCx) [inline src/server.ts]: we should replace this with running subscribe, also pass subscribe (optional) to this funtion.
only if subscribe is not defined, throw this error

File: src/server.ts [hunk 4]
```diff
@@ -105,32 +96,81 @@ class ExecuteAdapters {
...
         return {
-          unsubscribe: () => { /* Promises cannot be canceled */ },
+          unsubscribe: () => { /* Promises cannot be canceled */
+          },
         };
       },
     });
   }
 
+  public static executeFromAsyncIteratorFunction(handlerFn: Function, execute: Function): ExecuteReactiveFunction {
```
Comment (DxCx) [inline src/server.ts]: and this is how to call subscribe and convert to observable :)

File: src/server.ts [hunk 5]
```diff
@@ -272,12 +314,13 @@ export class SubscriptionServer {
 
     this.schema = schema;
     this.rootValue = rootValue;
-    if ( subscriptionManager ) {
+
+    if (subscriptionManager) {
       this.execute = ExecuteAdapters.executeFromSubscriptionManager(subscriptionManager);
-    } else if ( executor.executeReactive ) {
-      this.execute = executor.executeReactive.bind(executor);
+    } else if (subscribe) {
```
Comment (DxCx) [inline src/server.ts]: executeReactive shouldn't be removed.. :\

**Comment by mistic** [comment]
@Urigo I think u can close this PR. I've pushed your changes into my PR accordingly with @DxCx comments :)

**Pull request closed without merging**